kernel/sched.c

```diff
@@ -767,14 +767,44 @@ bool z_thread_prio_set(struct k_thread *thread, int prio)
 		need_sched = z_is_thread_ready(thread);
 
 		if (need_sched) {
+#if defined(CONFIG_SMP)
+			int  old_prio;
+
+			old_prio = thread->base.prio;
+#endif /* CONFIG_SMP */
 			/* Don't requeue on SMP if it's the running thread */
 			if (!IS_ENABLED(CONFIG_SMP) || z_is_thread_queued(thread)) {
 				dequeue_thread(thread);
 				thread->base.prio = prio;
 				queue_thread(thread);
+#if defined(CONFIG_SMP)
+				/*
+				 * IPIs may be needed when raising the priority
+				 * of a queued thread.
+				 */
+
+				if (old_prio > prio) {
+					flag_ipi();
+				}
+#endif /* CONFIG_SMP */
 			} else {
 				thread->base.prio = prio;
+#if defined(CONFIG_SMP)
+				/*
+				 * An IPI may be needed when lowering the
+				 * priority of a thread that is active on
+				 * another CPU.
+				 */
+
+				struct _cpu *cpu;
+
+				cpu = thread_active_elsewhere(thread);
+				if ((cpu != NULL) && (old_prio < prio)) {
+					flag_ipi();
```

Comment:
This is the kind of thing that creeps me out about changes like this.  This is going to iterate (!!) over all the CPUs in the scheduler every time any thread anywhere changes priority, putting a cost on all those small operations, just to "optimize" the case where you're "dropping the priority of a thread that happens to be running on another CPU".  Which is a crazy rare edge case, right?

This seems like exactly the circumstance where you'd want the simpler code and a broadcast IPI, no?  All these extra cycles add up over time, and suddenly our lean scheduler isn't so lean.

Reply:
Good point. I got so wrapped up in trimming out IPIs I forgot the larger picture.

Reply:
I've been thinking about this one some more, and the more I think about it, the more I really do think that it is beneficial. I think we are both in agreement that IPIs are needed on the "reschedule is needed" case. Where we primarily differ is the cost/benefit of the IPI calcuation.

While it is true that the cost for the CPU calling z_thread_prio_set() increases slightly, the total cost across the system decreases. That is, we can improve the overall performance at the cost of a slight increase in complexity.

There are three levels of calculations to consider (N shall be the number of CPUs) ...
1. When broadcasting, the total cost is  ...
`total average cost = (N - 1) * (time to service IPI)`
2. When conditionally broadcasting ...
`total average cost = (time to loop over N items) + (probability of crazy rare edge case) * (N - 1) * (time to service IPI)`
3. When targeting, the total average cost is ...
`total average cost = (time to loop over N items) + (probability of crazy rare edge case) * (time to service IPI)`

It is reasonable to expect that the "time to loop over N items" is going to be much lower than the time to service one interrupt (let alone N-1). Furthermore, as the probability decreases, so too does the contribution of the IPI servicing.